File: 0148-sort-list/0148-sort-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def sortList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        def merge(a, b):
            if not a:
                return b
            if not b:
                return a
            if a.val < b.val:
                a.next = merge(a.next, b)
                return a
            
            b.next = merge(a, b.next)
            return b
          
          
        if not head or not head.next:
            return head
        
        prev, slow, fast = None, head, head
        
        while fast and fast.next:
            prev = slow
            slow = slow.next
            fast = fast.next.next
        
        if prev:
            prev.next = None
            
        first = self.sortList(head)
        second = self.sortList(slow)
        
        return merge(first, second)

File: 0148-sort-list/0148-sort-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var sortList = function(head) {
  if (!head || !head.next) return head;
  
  let [prev, slow, fast] = [null, head, head];
  
  while (fast && fast.next) {
    prev = slow;
    slow = slow.next;
    fast = fast.next.next;
  }
  
  if (prev) prev.next = null;
  
  const first = sortList(head);
  const second = sortList(slow);
  
  return merge(first, second);
};

const merge = (a, b) => {
  if (!a) return b;
  if (!b) return a;
  
  if (a.val < b.val) {
    a.next = merge(a.next, b);
    return a;
  }
  
  b.next = merge(a, b.next)
  return b;
}